[BE] Add filter & sortings for the list of jobs

MihaiDruta opened the pull request:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Adding filters and sorting for activity list.

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Adding filters and sorting for activity list.

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Adding filters and sorting for activity list.

SpartakusMd commented on src/Filters/ActivityListFilter.php:
```diff
@@ -0,0 +1,76 @@
+    public function setFilterFields($filter)
+    {
+        if ($filter) {
+            if (array_key_exists('name', $filter)) {
```
Use `if (!empty('name', $filter))`. Because one could send empty parameters.

SpartakusMd commented on src/Filters/ActivityListFilter.php:
```diff
@@ -0,0 +1,76 @@
+        $this->userRepository = $userRepository;
+    }
+
+    public function setFilterFields($filter)
```
Parameter typehint and method return typehint.

SpartakusMd commented on src/Filters/ActivityListFilter.php:
```diff
@@ -0,0 +1,76 @@
+                $this->status = $filter['status'];
+            }
+            if (array_key_exists('owner', $filter)) {
+                $this->owner = $this->userRepository->find($filter['owner']);
```
Store the id, don't search it in the database. In the query builder you will do the search.

SpartakusMd commented on src/Filters/ActivityListFilter.php:
```diff
@@ -0,0 +1,76 @@
+                $this->activityType = $filter['activityType'];
+            }
+            if (array_key_exists('assignedUser', $filter)) {
+                $this->assignedUser = $this->userRepository->find($filter['assignedUser']);
```
Store the id, don't search it in the database.

SpartakusMd commented on src/Filters/ActivityListSort.php:
```diff
@@ -0,0 +1,34 @@
+    public function setSortingFields($sorting)
+    {
+        if ($sorting) {
+            if (array_key_exists('name', $sorting)) {
```
Use if (!empty('name', $filter)). Because one could send empty parameters.

SpartakusMd commented on src/Filters/ActivityListSort.php:
```diff
@@ -0,0 +1,34 @@
+    {
+        if ($sorting) {
+            if (array_key_exists('name', $sorting)) {
+                $this->name = $sorting['name'];
```
Validate what value is coming here. What if someone make a request `?sortby[name]=invalid_value` ?

SpartakusMd commented on src/Filters/ActivityListFilter.php:
```diff
@@ -0,0 +1,76 @@
+                $this->name = $filter['name'];
+            }
+            if (array_key_exists('status', $filter)) {
+                $this->status = $filter['status'];
```
Make sure status is integer.

SpartakusMd commented on src/Repository/ActivityRepository.php:
```diff
@@ -52,26 +55,72 @@ public function save(Activity $activity): void
-            ->select('activity')
+            ->select('DISTINCT activity')
             ->leftJoin('activity.activityUsers', 'activityUsers')
             ->where('activity.public = 1')
```
For these 3 `where/orWhere` you should be using https://www.doctrine-project.org/api/orm/latest/Doctrine/ORM/Query/Expr/Orx.html in order to exclude collision with the next conditions.

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Replacing array_key_exists() function with !empty($filter['field'])

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Parameter type and return type for setFilterFields() and setSortingFields()

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Parameter type and return type for setFilterFields() and setSortingFields()

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Validating int parameters for filtering and special parameters for sorting (asc, desc)

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Pagination is now a dynamic parameter like filters and sorting. Added sorting by seniority for the list of applicants.

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Fixed documentation on activity list and applicants list.

SpartakusMd commented on src/Filters/ActivityListPagination.php:
```diff
@@ -0,0 +1,32 @@
+        if (!empty($pagination['page']) && filter_var($pagination['page'], FILTER_VALIDATE_INT)) {
+            $this->currentPage = (integer)$pagination['page'];
+        } else {
+            $this->currentPage = 1;
```
FYI, if you declare the field with a default value `public $currentPage = 1;` you won't need the `else` branch ))

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Fixed documentation on activity list and applicants list.

Commit:
[BE] Add filter & sortings for the list of jobs
https://trello.com/c/T12F817f/52-be-add-filter-sortings-for-the-list-of-jobs

Fixed documentation on activity list and applicants list.
